morango/models.py

```diff
@@ -301,16 +301,18 @@ class Meta:
         unique_together = ("instance_id", "partition")
 
     @classmethod
-    def update_fsics(cls, fsic1, fsic2, sync_filter):
+    @transaction.atomic
+    def update_fsics(cls, fsics, sync_filter):
+        internal_fsic = DatabaseMaxCounter.calculate_filter_max_counters(sync_filter)
         updated_fsic = {}
-        for key, value in iteritems(fsic1):
-            if key in fsic2:
+        for key, value in iteritems(fsics):
+            if key in internal_fsic:
                 # if same instance id, update fsic with larger value
-                if fsic1[key] > fsic2[key]:
-                    updated_fsic[key] = fsic2[key]
+                if fsics[key] > internal_fsic[key]:
+                    updated_fsic[key] = fsics[key]
```

Comment:
I'm still wrapping my head around this approach, and whether there might be any issues. Here's where my head's at:

1. When we calculate the `internal_fsic`, some of the values might come from superset filters (rather than the filter we're using here itself).
2. This means that we might skip setting a DMC record for the current filter even if one doesn't exist for some instance_id at all.
3. This is a slightly different way of thinking about the DMC than we'd previously discussed, but I'm having trouble thinking of any way it would actually negatively impact behavior. If anything, it just keeps the DMC size down and should help performance, by not storing redundant entries in the DMC. Only negative impact I can imagine is that it makes it harder to reason about the state of the DMC if something *does* go wrong.

Do you see what I mean? Can you think of any possible issues with skipping saving these entries, as long as there's a superset filter with a higher value for that instance?

Reply:
Hard to see what could go wrong. I like to think of examples.
If we are syncing individual users at a time for a facility. Then we have filter, instance_id pairs for each user individually. Then we sync the whole facility, and it basically makes those other filters useless as we have the superset filter with a higher counter.
If we sync the whole facility first, and then do an individual user sync, the individual user filter counter will be higher. But if we re-sync the whole facility then that counter will be higher making the subset filter useless.